### PR TITLE
TMVA MultiProcessing

### PR DESCRIPTION
**Pull request by mammadhajili**
Mulltiprocessing of Hyper Parameter Optimisation, Cross Validation and Variable Importance

### CODE REVIEW

**Comment by lmoneta** [comment]
@phsft-bot build

**Comment by lmoneta** [comment]
Hi Mammad, 

 Can you please update your Pull request with the fixes for the formatting (clang-format) and also 
ass a tutorial program showing the usage of these 3 features in parallel.

 Thank you 

 Lorenzo

**Comment by martinmine** [comment]
@phsft-bot build

File: tmva/tmva/src/CrossValidation.cxx [hunk 1]
```diff
@@ -159,24 +162,45 @@ void TMVA::CrossValidation::Evaluate()
...
+  vector<pair<double, UInt_t>> res;
+
+  auto nWorkers = TMVA::gConfig().NWorkers();
+
+  if(nWorkers > 1) {
+      ROOT::TProcessExecutor workers(nWorkers);
+      res = workers.Map(workItem, ROOT::TSeqI(fNumFolds));
+  }
+
+  else {
```
Comment (ashlaban) [inline tmva/tmva/src/CrossValidation.cxx]: Use the structure
```
if () {
   ...
} else {
   ...
}

File: tmva/tmva/src/DataLoader.cxx [hunk 2]
```diff
@@ -853,17 +853,87 @@ TMVA::DataLoader* TMVA::DataLoader::MakeCopy(TString name)
...
+		  des->AddSignalTree((*treeinfo).GetTree(), (*treeinfo).GetWeight(),(*treeinfo).GetTreeType());
    }
 
    for( std::vector<TreeInfo>::const_iterator treeinfo=src->DataInput().Bbegin();treeinfo!=src->DataInput().Bend();treeinfo++)
    {
-      des->AddBackgroundTree( (*treeinfo).GetTree(), (*treeinfo).GetWeight(),(*treeinfo).GetTreeType());
+		  des->AddBackgroundTree((*treeinfo).GetTree(), (*treeinfo).GetWeight(),(*treeinfo).GetTreeType());
    }
 }
+std::vector<std::shared_ptr<TFile>> TMVA::DataLoaderCopyMP(TMVA::DataLoader *des, TMVA::DataLoader *src)
```
Comment (ashlaban) [inline tmva/tmva/src/DataLoader.cxx]: Add some documentation on what the method does, information about assumptions if there are any etc.

File: tmva/tmva/src/DataLoader.cxx [hunk 3]
```diff
@@ -853,17 +853,87 @@ TMVA::DataLoader* TMVA::DataLoader::MakeCopy(TString name)
...
+    else {
+			bfile = sfile;
+		}
+    TTree* signalTree = (TTree*)sfile->Get(stree->GetName());
+		TTree* backgTree = (TTree*)bfile->Get(btree->GetName());
+    des->AddSignalTree(signalTree);
+    des->AddBackgroundTree(backgTree);
+
+    vec_files.push_back(sfile);
+    vec_files.push_back(bfile);
```
Comment (ashlaban) [inline tmva/tmva/src/DataLoader.cxx]: Can these commented out lines be removed?

File: tmva/tmva/inc/TMVA/DataLoader.h [hunk 4]
```diff
@@ -226,7 +226,8 @@ namespace TMVA {
       ClassDef(DataLoader,3);
    };
    void DataLoaderCopy(TMVA::DataLoader* des, TMVA::DataLoader* src);
+   std::vector<std::shared_ptr<TFile>> DataLoaderCopyMP(TMVA::DataLoader *des, TMVA::DataLoader *src);
```
Comment (ashlaban) [inline tmva/tmva/inc/TMVA/DataLoader.h]: Do these need to be public? Would it be better to use private with Envelope as a friend? I think the intention is that only the Envelope should be able to use it. I would then propose that we enforce this. We can always open the interface up later, but not the other way around.

**Comment by ashlaban** [comment]
@phsft-bot build!

**Comment by ashlaban** [comment]
@phsft-bot build!

**Comment by martinmine** [comment]
@phsft-bot build!

**Comment by martinmine** [comment]
You might want to squash or fixup your clang-format commits.

**Comment by martinmine** [comment]
@phsft-bot build

**Comment by martinmine** [comment]
@phsft-bot build

**Comment by bellenot** [comment]
@lmoneta can you take care of this (or close it)

File: tmva/tmva/inc/TMVA/DataLoader.h (hunk 4 above)
Comment (etejedor) [reply tmva/tmva/inc/TMVA/DataLoader.h]: @mammadhajili could you answer the comments of @ashlaban please?

**Comment by etejedor** [comment]
@lmoneta @ashlaban all checks pass but there are some comments to be addressed, could you make sure this is followed or close the PR if not needed anymore?

**Comment by ashlaban** [comment]
Sorry for the late response. Yes, we'll make sure to follow up on this PR.

**Comment by etejedor** [comment]
@lmoneta @ashlaban any progress on merging this PR since our last chat?

**Comment by etejedor** [comment]
@lmoneta @ashlaban any progress on merging this PR since our last chat?

**Comment by ashlaban** [comment]
Cross validation has been integrated through a different PR. Remaining are the Hyper param opt and variable importance.
I refer to @lmoneta for the future plan there.

It is my understanding however that if we integrate them we do it through new PR’s, indicating that this PR can be closed.

**Comment by ashlaban** [comment]
These changes will be implemented in separate PR's.

**Pull request closed without merging**